### PR TITLE
Speed up Code that Checks Optical Duplicates.

### LINKED ISSUE #1325: Samtools markdup with high depth regions
#### Are you using the latest version of [samtools](https://github.com/samtools/samtools/releases/latest) and [HTSlib](https://github.com/samtools/htslib/releases/latest)? If not, please specify.
(run `samtools --version`)

I am using samtools 1.11 (htslib 1.11)

 #### Please describe your environment.
- OS (run `uname -sr` on Linux/Mac OS or ` wmic os get Caption, Version` on Windows)
- machine architecture (run `uname -m` on Linux/Mac OS or ` wmic os get OSArchitecture` on Windows)
- compiler (run `gcc --version` or `clang --version`)

CentOS Linux 7 x86_64 with gcc 4.8.5

#### Please specify the steps taken to generate the issue, the command you are running and the relevant output.
I have encountered an issue with samtools markdup running past 8 days without actually completing on a few BAMs. I don’t get any relevant output since it just doesn’t finish. Digging deeper, I found out that a ~400 bp region in chromosome 2, with more than 8 million reads, is causing this issue. It looks like markdup just isn't optimized to deal with abnormally high depth regions, unless, I missed some parameter that would help. I have created micro BAMs of that region (with an additional 500bp on both ends) and tried running markdup on them but I still get the same results. For both types of BAMs, I used the command (BAM names were replaced with "micro"):
```
samtools markdup \
    -d “2500” \
    -S \
    -f “micro.samtools.markdup.txt” \
    --threads 4 \
    --write-index \
    -T “micro” \
    “micro.bam” \
    “micro.md.bam”
```
Also, I ran GATK MarkDuplicates (v4.1.8.0) on these micro BAMs and they finished under 35 minutes. The GATK command I used was the following:
```
gatk MarkDuplicates \
    --java-options “-Xmx28G” \
    --TMP_DIR “temp” \
    --ASSUME_SORT_ORDER coordinate \
    --CLEAR_DT false \
    --ADD_PG_TAG_TO_READS false \
    --OPTICAL_DUPLICATE_PIXEL_DISTANCE “2500” \
    --INPUT "micro.bam" \
    --METRICS_FILE “micro.gatk.mdmetrics.txt” \
    --OUTPUT “micro.md.bam”
```
If needed, I could provide a micro BAM that is causing this issue. Thank you.

### PR DESCRIPTION
**Pull request by whitwham**
Duplicates can appear in any order in a bam file so some degree of backtracking is needed to correctly identify if a duplicate is actually an optical one. The existing code did not scale when a single read can have tens or hundreds of thousands of duplicates.

This should fix #1325, or at least ameliorate it.

### CODE REVIEW

**Comment by daviesrob** [comment]
I'm getting a use-after-free error from address sanitizer - it looks like a stale bam record is on the `ori` list [here](https://github.com/samtools/samtools/blob/b965b42c/bam_markdup.c#L1063).  The test harness doesn't show this, so probably isn't doing enough to find the bug.  I'll point you at my input file so you can investigate.

**Comment by whitwham** [comment]
> I'm getting a use-after-free error from address sanitizer - it looks like a stale bam record is on the `ori` list [here](https://github.com/samtools/samtools/blob/b965b42c/bam_markdup.c#L1063). The test harness doesn't show this, so probably isn't doing enough to find the bug. I'll point you at my input file so you can investigate.

Thanks.  I will take a look using your test data.

File: bam_markdup.c [hunk 1]
```diff
@@ -943,118 +1086,227 @@ static int duplicate_chain_check(md_param_t *param, khash_t(duplicates) *dup_has
...
+                break;
+            }
+        } else {
+            // this is the last set of results and the end entry will be blank
+            if (!bam_get_qname(in_read->b)) {
+                break;
+            }
+        }
+
+        if (!(in_read->b->core.flag & BAM_FDUP) && in_read->duplicate) { // is the head of a duplicate chain
```
Comment (jkbonfield) [inline bam_markdup.c:1279]: On a small test file it still seems to spend longer than I expected, with about 50% of time at this line of code surprisingly.

It got about 20% quicker just by swapping the order of conditionals, ie:

```
if (in_read->duplicate && !(in_read->b->core.flag & BAM_FDUP))
```

I'm not sure if my test data was abnormal in some way though.

**Review by jkbonfield — COMMENTED** [review]
Only very minor code comments.

I'm struggling to get it to work well though on a small test set.  Maybe it's something wrong with how I prepared my data.  I'll email you the details.

File: bam_markdup.c (hunk 1 above)
Comment (jkbonfield) [reply bam_markdup.c:1279]: Ignore this.  While still more efficient the other way around, I think it *shouldn't* be spending so long in this function.  It's a symptom of my input file having already marked for duplicates and not using the "-c" option to clear those flags.

I don't know yet why that makes such a difference though.

File: bam_markdup.c [hunk 2]
```diff
@@ -693,6 +711,58 @@ static inline int get_coordinate_positions(const char *qname, int *xpos, int *yp
...
+
+static int get_coordinates(const char *name, int *xpos_out, long *x_coord, long *y_coord, long *warnings) {
+    int ret = 1;
+    int seps, xpos = 0, ypos = 0;
+    long x = 0, y = 0;
+    char *end;
+
+    seps = get_coordinate_positions(name, &xpos, &ypos);
+
+    if (!(seps == 3 || seps == 4 || seps == 6 || seps == 7)) {
```
[the remark this answers is not in the document]
Comment (jkbonfield) [reply bam_markdup.c:731]: Thanks for the commenting.  I guess the answer to "_7th element is it's not considered in that step. Does that matter?_" is "maybe".  That'll be a different issue if done though.  (See #1358)

[the remark this answers is not in the document]
Comment (whitwham) [reply bam_markdup.c:731]: It doesn't matter **yet**.

**Comment by jkbonfield** [comment]
Thanks for the updates.

I can confirm it now works much faster even on data that's already had markdups run before.  However it now appears to duplicate some tag entries.  Eg after a 2nd run:

```
seq     1168    1       9996    16      112S39M *       0       0       <seq> <qual> NM:i:0  MD:Z:39 AS:i:39 XS:i:35 XA:Z:6,-147869,116S35M,0;hs37d5,+10061146,35M116S,0;7,-10002,117S34M,0;15,+102521304,34M117S,0;2,+241161288,34M117S,0;  do:Z:seq2    dt:Z:LB do:Z:seq2    dt:Z:LB
```

Note `do` and `dt` tags are doubled up.  Changing all the `bam_aux_append` calls with `bam_aux_update_str` seems to cure the problem and I cannot see any impact on the time taken.  1 million seqs with append vs update (uncompressed BAM output) took 4.60s vs 4.65s.  The second pass where update actuallys updates instead of appending took 4.35s, so it's now faster (fewer reallocs as a guess).  The md5sum output of 1st and 2nd pass is identical, as is 1st past between append vs update.

My diff:

```
diff --git a/bam_markdup.c b/bam_markdup.c
index 10a3f66..2da184f 100644
--- a/bam_markdup.c
+++ b/bam_markdup.c
@@ -945,7 +945,7 @@ static int mark_duplicates(md_param_t *param, khash_t(duplicates) *dup_hash, bam
     dup->core.flag |= BAM_FDUP;
 
     if (param->tag) {
-        if (bam_aux_append(dup, "do", 'Z', strlen(bam_get_qname(ori)) + 1, (uint8_t*)bam_get_qname(ori))) {
+        if (bam_aux_update_str(dup, "do", strlen(bam_get_qname(ori)) + 1, bam_get_qname(ori))) {
             fprintf(stderr, "[markdup] error: unable to append 'do' tag.\n");
             return -1;
         }
@@ -953,12 +953,12 @@ static int mark_duplicates(md_param_t *param, khash_t(duplicates) *dup_hash, bam
 
     if (param->opt_dist) { // mark optical duplicates
         if (optical_duplicate(ori, dup, param->opt_dist, warn)) {
-            bam_aux_append(dup, "dt", 'Z', 3, (const uint8_t *)"SQ");
+            bam_aux_update_str(dup, "dt", 3, "SQ");
             dup_type = 'O';
             (*optical)++;
         } else {
             // not an optical duplicate
-            bam_aux_append(dup, "dt", 'Z', 3, (const uint8_t *)"LB");
+            bam_aux_update_str(dup, "dt", 3, "LB");
         }
     }
 
@@ -1909,7 +1909,7 @@ static int bam_mark_duplicates(md_param_t *param) {
                     np_duplicate++;
 
                     if (param->tag && kh_val(dup_hash, k).name) {
-                        if (bam_aux_append(b, "do", 'Z', strlen(kh_val(dup_hash, k).name) + 1, (uint8_t*)kh_val(dup_hash, k).name)) {
+                        if (bam_aux_update_str(b, "do", strlen(kh_val(dup_hash, k).name) + 1, (char*)kh_val(dup_hash, k).name)) {
                             fprintf(stderr, "[markdup] error: unable to append supplementary 'do' tag.\n");
                             goto fail;
                         }
@@ -1917,10 +1917,10 @@ static int bam_mark_duplicates(md_param_t *param) {
 
                     if (param->opt_dist) {
                         if (kh_val(dup_hash, k).type) {
-                            bam_aux_append(b, "dt", 'Z', 3, (const uint8_t *)"SQ");
+                            bam_aux_update_str(b, "dt", 3, "SQ");
                             np_opt_duplicate++;
                         } else {
-                            bam_aux_append(b, "dt", 'Z', 3, (const uint8_t *)"LB");
+                            bam_aux_update_str(b, "dt", 3, "LB");
                         }
                     }
                 }
```

If you wish I can apply this and squash it into the other commits during merge.

Assuming you're happy for the various commits to get squashed together?

**Comment by whitwham** [comment]
You can.  I guess this is more fallout from not clearing the previously dup marking data.  That really should have been the default action.

**Comment by jkbonfield** [comment]
Ok will do.

**Pull request merged**